src/core/DY.Core.Trace.SplitUtils.fst

```diff
@@ -0,0 +1,306 @@
+    (ensures trace_entry_equiv e2 e1)
+  = ()
+
+let trace_entry_equiv_transitive (#label_t:Type) (e1 e2 e3:trace_event_ label_t)
+  : Lemma
+    (requires
+       trace_entry_equiv e1 e2 /\
+       trace_entry_equiv e2 e3
+    )
+    (ensures trace_entry_equiv e1 e3)
+  = ()
+
+(*** Trace search functions ***)
+
+/// As trace entries are not an eqtype, rather than looking up a specific trace entry,
+/// we first implement a function that lets us look up trace entries with a given property.
+/// Currently, this will find the most recent entry satisfying the property, if multiple
+/// exist, but it is easy to generalize to find the oldest such entry, or all such entries.
+///
+/// Using trace entry equivalence and this first search function, we can find entries that are
+/// equivalent to a known trace entry, though they may not necessarily be equal.
+
+let rec trace_findP (#label_t:Type) (tr:trace_ label_t) (p:trace_event_ label_t -> bool)
+  : Pure (option timestamp)
+    (requires True)
+    (ensures fun ts_opt ->
+       match ts_opt with
+       | None -> forall ts. ts < length tr ==> ~(p(get_event_at tr ts))
+       | Some ts -> ts < length tr /\ p (get_event_at tr ts)
+    )
+  = match tr with
+    | Nil -> None
+    | Snoc hd entry ->
+      if p entry then
+        Some (last_ts tr)
+      else
+        trace_findP hd p
+
+let trace_find (#label_t:Type) (tr:trace_ label_t) (e:trace_event_ label_t{event_exists tr e})
+  : Pure timestamp (requires True) (ensures fun ts -> ts < length tr /\ trace_entry_equiv e (get_event_at tr ts))
+  = let Some ts = trace_findP tr (fun e' -> trace_entry_equiv e e') in
+    ts
+
+
+(*** Trace splitting functions ***)
+
+/// In reasoning about the trace, we most often make use of the prefix function to extract
+/// the portion of a trace before some entry that we are focused on. For some purposes, however,
+/// for instance, to give a modifies analysis for a trace-passing function which specifies what
+/// pieces of state it may modify, it is useful to look at suffixes as well (e.g., the suffix of
+/// the output trace that is its "difference" with the input trace).
+///
+/// The trace splitting functions here unify getting prefixes and suffixes, either based on a
+/// timestamp or a particular event.
+///
+/// The trace split function is also compatible with the various other trace functions ---
+/// in particular, length and get_event_at, which give us a correctness guarantee of the split,
+/// but also prefix, grows/<$, and fmap_trace.
+
+let rec trace_split (#label_t:Type) (tr:trace_ label_t) (ts:timestamp{ts < length tr})
+  : (trace_ label_t & trace_event_ label_t & trace_ label_t)
+  = let Snoc hd entry = tr in
+    if ts = last_ts tr then
+      (hd, entry, Nil)
+    else
+      let (tr1, e, tr2') = trace_split hd ts in
+      (tr1, e, Snoc tr2' entry)
+
+let rec trace_split_length (#label_t:Type) (tr:trace_ label_t) (ts:timestamp{ts < length tr})
+  : Lemma
+    (requires True)
```

Comment:
Isn't it unnecessary to specify `requires True`? Applies also to the following lemmas.

Reply:
Yes, you're right --- I think those came out of some older `Pure` functions, which do need a `requires`.